services/code_hosting/git_hub_service.rb

```diff
@@ -1,118 +1,352 @@
+      # @return [Git::Base]
+      def clone(repo_url: nil, branch: "master", provider_credential: nil, path: self.temp_path)
+        repo = self.repo_from_url(repo_url)
+        path = File.join(path, repo, branch)
+        FileUtils.rm_rf(path) if File.directory?(path)
+        FileUtils.mkdir_p(path)
+        self.setup_auth(repo_url: repo_url, provider_credential: provider_credential, path: path)
+        Git.clone(url_from_repo(repo_url), repo.split("/").last,
+                  path: path,
+                  recursive: true,
+                  depth: 1)
+        git = Git.open(File.join(path, repo.split("/").last))
+        git.branch(branch).checkout
+        return git
+      end
 
-    GitHubService.status_context_prefix = "fastlane.ci: "
+      # Class method that finds the directory for the first Fastfile found given a root_path
+      # @param [String] root_path
+      # @return [String] the path of the Fastfile
+      def fastfile_path(root_path: nil)
+        fastfiles = Dir[File.join(root_path, "fastlane/Fastfile")]
+        fastfiles = Dir[File.join(root_path, "**/fastlane/Fastfile")] if fastfiles.count == 0
+        fastfile_path = fastfiles&.first
+        return fastfile_path
+      end
+
+      # Class method that setups the authentication needed for making git operations.
+      # @param [String] repo_url
+      # @param [GithubProviderCredential] provider_credential
+      # @param [String] path
+      def setup_auth(repo_url: nil, provider_credential: nil, path: nil)
+        repo = repo_from_url(repo_url)
+        self.temporary_storage_path = File.join(self.temporary_git_storage, "git-auth-#{SecureRandom.uuid}")
+        # More details: https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage
+
+        FileUtils.mkdir_p(path) unless File.directory?(path)
+
+        store_credentials_command = "git credential-store --file #{self.temporary_storage_path.shellescape} store"
+        content = [
+          "protocol=https",
+          "host=#{provider_credential.remote_host}",
+          "username=#{provider_credential.email}",
+          "password=#{provider_credential.api_token}",
+          ""
+        ].join("\n")
+
+        scope = "local"
+
+        unless File.directory?(File.join(path, repo.split("/").last, ".git"))
+          # we don't have a git repo yet, we have no choice
+          # TODO: check if we find a better way for the initial clone to work without setting system global state
+          scope = "global"
+        end
+        use_credentials_command = "git config --#{scope} credential.helper 'store --file #{self.temporary_storage_path.shellescape}' #{File.join(path, repo.split('/').last)}"
+
+        cmd = TTY::Command.new(printer: :quiet)
+        cmd.run(store_credentials_command, input: content)
+        cmd.run(use_credentials_command)
+      end
+
+      # Class method that removes the authentication stored for git operations.
+      def unset_auth
```

Comment:
Maybe here I should to inject the random UUID generated in order to cleanup the directory between executions

Reply:
I don't know that is needs to be random. It could just be a unique hash of org, repo name, and committer

Reply:
Yeah, I meant, as you have commited in #290, I should be able to pass that `id` into this method to cleanup the directory (as for now it doesn't know what he is actually deleting)